apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+  graphs easier to read.
+
+### Capsule Library Rules ###
+
+The rules for capsule libraries are also straightforward:
+
+- Do not depend on feature libraries
+- Do not read from or write to the global window object (accessing standard browser
+  APIs is OK)
+- Do not reference DOM nodes by ID or css selectors that are not passed in to
+  the library explicitly.
+
+
+## Feature Library ##
+
+Feature libraries require more discipline to avoid coupling and spaghetti code,
+since they must inherently interact with some form of global state at some
+point. But if we are careful and consistently apply a sensible pattern for
+writing feature libraries, we can keep things clean and manageable.
+
+### Feature Library Pattern ###
+
+The primary goal of the feature library pattern is to make it really clear where
+and how the library interacts with shared state or remote services.
+
+#### Exposing/Providing Shared State ####
+
+Shared application state should be managed by redux. Since there can only be one
+redux store on a page, while there can be multiple features on a page, feature
+libraries are not allowed to create their own redux store. However, they can
+have their own actions and reducers which operate on a well defined portion of
+the global redux state tree. https://github.com/erikras/ducks-modular-redux
+describes one pattern for doing this, which we will riff on.
+
+Feature libraries that expose shared redux state should have the following
+directory structure:
+
+```
+src/
+  <feature-name>/
+    index.js
+    redux/
+      reducer.js
+      actionTypes.js
+      actionCreators.js
+      selectors.js
+```
+
+In this hypothetical feature, we expose the state and actions for a stepper
+feature for use in other libraries. In order to keep encapsulation high, we want
+to only expose the action types, action creators, and selectors that other
+libraries need to use, while keeping the rest more "private" to our own
+library. Also, it's important to consider that other code which uses this feature
+library should not have to know about the shape of the state provided by the
+feature. Likewise, the feature should not have to know how it's state gets
+stored in the global redux state tree.
+
+Here is how the above files might look if we wanted to expose only step
+incrementation, but not step decrementation.
+
+**`src/stepper/redux/actionTypes.js`**
+
+Action types are prefixed with the name of the feature to avoid collisions with
+other libraries and their reducers.
+
+```js
+export const INCREMENT_STEP = "stepper/INCREMENT_STEP";
+export const DECREMENT_STEP = "stepper/DECREMENT_STEP";
+```
+
+**`src/stepper/redux/reducer.js`**
+
+The reducer for the feature library should operate on a subtree of the global
+state tree, rather than on the entire state tree. Note that this library acts
+like a singleton in the sense that there is only one corresponding state object
+for the library on the page at any given time.
```

Comment:
Shouldn't need/want them when writing redux, which very much encourages use of pure functions.